packages/subgraph/src/utils/tokens/erc721.ts

```diff
@@ -0,0 +1,234 @@
+import {Address, BigInt, Bytes, ethereum, log} from '@graphprotocol/graph-ts';
+import {
+  ERC721Balance,
+  ERC721Contract,
+  ERC721Transfer
+} from '../../../generated/schema';
+import {ERC721} from '../../../generated/templates/DaoTemplate/ERC721';
+import {supportsInterface} from '../erc165';
+import {DECODE_OFFSET, getTransferId, TransferType} from './common';
+import {
+  ERC721_safeTransferFromNoData,
+  ERC721_safeTransferFromWithData,
+  ERC721_transferFrom
+} from './common';
+
+function supportsERC721(token: Address): bool {
+  // Double check that it's ERC721 by calling supportsInterface checks.
+  let erc721 = ERC721.bind(token);
+  let introspection_01ffc9a7 = supportsInterface(erc721, '01ffc9a7'); // ERC165
+  let introspection_80ac58cd = supportsInterface(erc721, '80ac58cd'); // ERC721
+  let introspection_00000000 = supportsInterface(erc721, '00000000', false);
+  return (
+    introspection_01ffc9a7 && introspection_80ac58cd && introspection_00000000
+  );
+}
+
+export function fetchERC721(address: Address): ERC721Contract | null {
+  let erc721 = ERC721.bind(address);
+
+  // Try load entry
+  let contract = ERC721Contract.load(address.toHexString());
+  if (contract != null) {
+    return contract;
+  }
+
+  // Detect using ERC165
+  if (!supportsERC721(address)) {
+    return null;
+  }
+
+  contract = new ERC721Contract(address.toHexString());
+
+  let try_name = erc721.try_name();
+  let try_symbol = erc721.try_symbol();
+  contract.name = try_name.reverted ? '' : try_name.value;
+  contract.symbol = try_symbol.reverted ? '' : try_symbol.value;
+
+  contract.save();
+
+  return contract;
+}
+
+export function updateERC721Balance(
+  daoId: string,
+  token: string,
+  tokenId: BigInt,
+  timestamp: BigInt,
+  type: TransferType
+): void {
+  let balanceId = daoId.concat('_').concat(token);
+  let erc721Balance = ERC721Balance.load(balanceId);
+
+  if (!erc721Balance) {
+    erc721Balance = new ERC721Balance(balanceId);
+    erc721Balance.dao = daoId;
+    erc721Balance.token = token;
+    erc721Balance.tokenIds = [];
+  }
+
+  let tokenIds = erc721Balance.tokenIds;
+  if (type == TransferType.Withdraw) {
+    tokenIds.splice(tokenIds.indexOf(tokenId), 1);
+  } else {
+    tokenIds.push(tokenId);
+  }
+
+  erc721Balance.tokenIds = tokenIds;
+  erc721Balance.lastUpdated = timestamp;
+  erc721Balance.save();
+}
+
+export function handleERC721Received(
+  token: Address,
+  dao: Address,
+  data: Bytes,
+  event: ethereum.Event
+): void {
+  let contract = fetchERC721(token);
+  if (!contract) {
+    return;
+  }
+
+  let calldata = DECODE_OFFSET + data.toHexString().slice(10); // TODO: Giorgi insted aof 0x, append DECODE_OFFSET
```

Comment:
do the TODO

Reply:
There was a comment remaining, it's actually done. So I removed the TODO comment.